internal/statemachine/common_states.go

```diff
@@ -133,66 +134,133 @@ func (stateMachine *StateMachine) generateDiskInfo() error {
+				"than requested image size (%d). Try using a larger value of "+
+				"--image-size",
+				stateMachine.RootfsSize, desiredSize,
+			)
+		} else {
 			stateMachine.RootfsSize = desiredSize
 		}
 	}
 
-	stateMachine.syncGadgetStructureRootfsSize()
+	if rootfsStructure != nil {
+		setStructureSize(rootfsStructure, stateMachine.RootfsSize)
+	}
+
 	return nil
 }
 
-// findRootfsVolume finds the volume associated to the rootfs
-func (stateMachine *StateMachine) findRootfsVolume() (*gadget.Volume, string) {
+// getRootfsMinSize gets the minimum size needed to hold the built rootfs directory
+func (stateMachine *StateMachine) getRootfsMinSize() (quantity.Size, error) {
+	rootfsMinSize, err := helper.Du(stateMachine.tempDirs.rootfs)
+	if err != nil {
+		return quantity.Size(0), fmt.Errorf("Error getting rootfs size: %s", err.Error())
+	}
+
+	// Take into account ext4 filesystems metadata size
+	rootfsPadding := ext4Padding
+	rootfsMinSize = quantity.Size(math.Ceil(float64(rootfsMinSize) * ext4FudgeFactor))
+	rootfsMinSize += rootfsPadding
+
+	return stateMachine.alignToSectorSize(rootfsMinSize), nil
+}
+
+// getGadgetRootfsMinSize gets the minimum size of the rootfs requested in the
+// gadget YAML
+func (stateMachine *StateMachine) getGadgetRootfsMinSize(rootfsStructure *gadget.VolumeStructure) quantity.Size {
+	if rootfsStructure == nil {
+		return quantity.Size(0)
+	}
+
+	return stateMachine.alignToSectorSize(rootfsStructure.MinSize)
+}
+
+// getRootfsDesiredSize subtracts the size and offsets of the existing structures
+// from the requested image size to determine the maximum possible size of the rootfs
+func (stateMachine *StateMachine) getRootfsDesiredSize(rootfsVolume *gadget.Volume, rootfsVolumeName string) (quantity.Size, bool) {
+	if rootfsVolume == nil {
+		return quantity.Size(0), false
+	}
+
+	desiredSize, found := stateMachine.ImageSizes[rootfsVolumeName]
+	if !found {
+		// So far we do not know of a desired size for the rootfs
+		return quantity.Size(0), found
+	}
+
+	reservedSize := calculateNoRootfsSize(rootfsVolume)
+	// Create a temp partition table to get its size
+	// The value is already a multiple of the sector size, so no need to align again
+	tempPartitionTable := partition.NewPartitionTable(rootfsVolume, uint64(stateMachine.SectorSize), 0)
+	reservedSize += quantity.Size(tempPartitionTable.PartitionTableSize())
+
+	desiredSize = helper.SafeQuantitySubtraction(desiredSize, reservedSize)
+	desiredSize = stateMachine.alignToSectorSize(desiredSize)
+
+	return desiredSize, found
+}
+
+// calculateNoRootfsSize determines the needed space for existing structures
+// except for the rootfs
+func calculateNoRootfsSize(v *gadget.Volume) quantity.Size {
+	var size quantity.Size
+	for _, s := range v.Structure {
+		if helper.IsRootfsStructure(&s) { //nolint:gosec,G301
+			continue
+		}
+		if s.Offset != nil && quantity.Size(*s.Offset) > size {
+			size = quantity.Size(*s.Offset)
+		}
+		size += s.MinSize
```

Comment:
It's a tangent again (sorry) but code like this makes some assumptions about the order of structures in the volume, like that they are in offset order, and that the rootfs structure is last. Does snapd validate these when parsing the gadget.yaml?

Reply:
later I found that snapd does order the structures by offset -- and ensures s.Offset is set. Which seems to imply that we don't need to be so careful here? (except probably for tests?)